### ISSUE TITLE
Tracking issue: Custom Elements Hydration 🧩

### ISSUE THREAD

**Issue by luisherranz** [WordPress/block-hydration-experiments#39]
This issue aims to list and keep track of the tasks related to the Custom Elements Hydration experiment. 

_For the Directives Hydration experiment, please check the [Tracking Issue: Directives Hydration ⚛️](https://github.com/WordPress/block-hydration-experiments/issues/64)._

The main branch of this experiment is `main-custom-elements-hydration`. Make sure you select it when opening a pull request if it is related to this experiment.

Please proactively comment on this issue whenever you start working on a new task, get blocked, or finish a task, sharing as much detail and information as possible. Thanks!!

# Done

- Hydrate Frontend components
- Make sure Frontend components are automatically rehydrated if they appear in the DOM at any point (not only on page load)
- Support partial hydration with Inner blocks (children raw HTML)
- Serialize attributes and pass them down to the View component
- Use children instead of in the Save component to be able to reuse the same component in the View
- https://github.com/WordPress/block-hydration-experiments/pull/15
- https://github.com/luisherranz/block-hydration-experiments/pull/8
- https://github.com/luisherranz/block-hydration-experiments/pull/3
- https://github.com/luisherranz/block-hydration-experiments/pull/2
- https://github.com/luisherranz/block-hydration-experiments/pull/3
- https://github.com/luisherranz/block-hydration-experiments/pull/14
- https://github.com/luisherranz/block-hydration-experiments/pull/7
- https://github.com/WordPress/block-hydration-experiments/pull/34
- https://github.com/WordPress/block-hydration-experiments/pull/30
- https://github.com/WordPress/block-hydration-experiments/pull/40
- https://github.com/WordPress/block-hydration-experiments/pull/28
- https://github.com/WordPress/block-hydration-experiments/pull/41
- https://github.com/WordPress/block-hydration-experiments/pull/45
- https://github.com/WordPress/block-hydration-experiments/pull/47
- https://github.com/WordPress/block-hydration-experiments/pull/57

# Required to expose as experimental API in Gutenberg

- https://github.com/WordPress/block-hydration-experiments/issues/49
- https://github.com/WordPress/block-hydration-experiments/issues/11
- https://github.com/WordPress/block-hydration-experiments/issues/50
- https://github.com/WordPress/block-hydration-experiments/issues/59
- https://github.com/WordPress/block-hydration-experiments/issues/69

# Future improvements

- https://github.com/WordPress/block-hydration-experiments/issues/62
- https://github.com/WordPress/block-hydration-experiments/issues/58
- Add support for frontend-side i18n
- Research how to hydrate different block variations
- Change hydration technique based on block attributes
- Add an onClick hydration technique (like [Qwik](https://github.com/BuilderIO/qwik) or [Rocket](https://next.rocket.modern-web.dev/docs/basics/hydration/))
- Support for Suspense and Error boundaries between different blocks 
- Search for alternative bundling strategies to `wp` globals that can support tree-shaking and code deduplication

**Comment by luisherranz** [comment]
# Previous video summaries

Originally posted in https://github.com/WordPress/block-hydration-experiments/issues/6

## Update 1

<a href="https://www.loom.com/share/549258dfe6844fe0b3fdc8c3fc856b13">
  <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/549258dfe6844fe0b3fdc8c3fc856b13-with-play.gif">
</a>

https://www.loom.com/share/549258dfe6844fe0b3fdc8c3fc856b13

For context, when I mention _Turbolinks_ in the video, I'm referring to this type of client-side navigation: https://github.com/WordPress/gutenberg/pull/38713 

## Update 2

<a href="https://www.loom.com/share/e5a5541d027d4149a91a70f30b0913be">
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/e5a5541d027d4149a91a70f30b0913be-with-play.gif">
  </a>

https://www.loom.com/share/e5a5541d027d4149a91a70f30b0913be

Resources mentioned:

- [Block Context](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-context/)
- [Attribute serialization](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/#value-source)
- [hpq](https://github.com/aduth/hpq)

**Comment by michalczaplinski** [comment]
I have started working on adding [support for providing context from non-interactive blocks](https://github.com/WordPress/block-hydration-experiments/pull/28#).

**Comment by michalczaplinski** [comment]
Resharing this brief [status update](https://github.com/WordPress/block-hydration-experiments/pull/28#issuecomment-1175616435) on my work on #28:

I was able to make it work so far in the simplest case: one static (non-interactive) parent and one child interactive block 👍 .

Most immediate next steps would be:
- Making it work with multiple parents. An interactive block can inherit context from multiple parent blocks.
- Making it work with initially hidden blocks. Adding a global event listener on load does not work because if a new interactive block is dynamically added to the DOM, it will not be subscribed to the events! I guess we have to encapsulate this code in the `connectedCallback()` of our `<gutenberg-interactive-block>` custom component.

**Comment by michalczaplinski** [comment]
I've finished the work on [Support providing context from non-interactive blocks #28](https://github.com/WordPress/block-hydration-experiments/pull/28). It's gonna go through some review now but should be close to what we want to merge 🙂 .

**Comment by c4rl0sbr4v0** [comment]
[The React Context API subtasks are 90% complete!](https://github.com/WordPress/block-hydration-experiments/pull/40#issuecomment-1174945660) We still need to check how it works on the editor, where it should be easy.

I will update the tracking issue again as soon as we unsubscribe the contexts after using them.

**Comment by luisherranz** [comment]
I've opened an issue to see how we could unmount the interactive-block components before they are removed from the DOM:

- https://github.com/WordPress/block-hydration-experiments/issues/42

**Comment by luisherranz** [comment]
I've also added Matías' nomenclature suggestions to the list, although I'd wait until the rest of the PRs are merged to avoid merge conflicts.

**Comment by c4rl0sbr4v0** [comment]
#40 Status update:

The last commits are adding a double Map to link blocks to context, and then the values and update values functions to those block->Context-> Value

On the other hand, I will next take a look if we can suppress custom warnings for the useEffect, or either way, look for an alternative to run the code once without it.

Weakmap seems to be valid in most edge cases, but @luisherranz offered to take a deeper view to see if it fits our requirements.

I will then wait for #28 to be merged, so we can check later if there are any conflicts and we can start testing altogether.

**Comment by michalczaplinski** [comment]
- #28 status update:

Luis has [suggested](https://github.com/WordPress/block-hydration-experiments/pull/28#discussion_r920274594) that we try a slightly different approach in that PR in order to avoid using a custom element to wrap all blocks that pass context. 

The idea is to serialize the context and pass it as an HTML attribute of the top-level HTML element of the block itself (as opposed to setting it on the custom element). Then, we can use a MutationObserver on the `document` to detect changes to the DOM and pass the `context` whenever that happens. This is necessary because the "new", updated DOM might contain new interactive blocks that are subscribed to some piece of `context`.

My worry is that this can have poor performance because we'll have to fire the MutationObserver callback on every DOM change and the DOM changes can come from unexpected sources (browser extensions, ads, third party scripts, etc.)

**Next step**: I will make a quick test to try to see how bad the new implementation would perform if there are very frequent DOM updates.

**Comment by c4rl0sbr4v0** [comment]
-https://github.com/WordPress/block-hydration-experiments/pull/40 status update:

I added another Context to see if it works and seems to be failing. Feel free to keep working on that, as I will be 2 weeks AFK!

**Comment by luisherranz** [comment]
Before we start adding code to the [WooCommerce Block Hydration Experiments](https://github.com/woocommerce/woocommerce-blocks-hydration-experiments/issues/20) repo, I'd like to finish the three PRs we have open right now:

- https://github.com/WordPress/block-hydration-experiments/pull/40 – @c4rl0sbr4v0 is on holiday so I can take care of this
- https://github.com/WordPress/block-hydration-experiments/pull/28
- https://github.com/WordPress/block-hydration-experiments/pull/41


And once those are merged, I'd like to do the nomenclature refactor:

- https://github.com/WordPress/block-hydration-experiments/issues/37

_Not important for Woo, but at that point, I'd also like to remove dprint, and test prettier with the [prettier-php](https://github.com/WordPress/block-hydration-experiments/issues/16#issuecomment-1190144963) plugin._

**Comment by luisherranz** [comment]
I've finished the support for the React Context API (https://github.com/WordPress/block-hydration-experiments/pull/40), including a code refactoring of `frontend.js`, which was getting a bit out of hand (https://github.com/WordPress/block-hydration-experiments/pull/40/commits/d61391cf240e82b5edafbd8f3af081f41331318d). I've also resolved the conflicts on https://github.com/WordPress/block-hydration-experiments/pull/28 and I'm working to resolve them in https://github.com/WordPress/block-hydration-experiments/pull/41.

**Comment by luisherranz** [comment]
I've switched to `prettier` in https://github.com/WordPress/block-hydration-experiments/pull/43 and resolved conflicts on https://github.com/WordPress/block-hydration-experiments/pull/28 and https://github.com/WordPress/block-hydration-experiments/pull/41.

@michalczaplinski, @ockham everything should be fine now.

**Comment by luisherranz** [comment]
https://github.com/WordPress/block-hydration-experiments/pull/28 should be ready now. [I've made a video](https://github.com/WordPress/block-hydration-experiments/pull/28#issuecomment-1194147957) to explain the changes because I unified everything in `gutenberg-block`.

I want to take a fresh look at the inner blocks template to see if we can avoid this problem now (and don't wait if/until we do this in Gutenberg):

- https://github.com/WordPress/block-hydration-experiments/issues/29

**Comment by luisherranz** [comment]
@ockham finished https://github.com/WordPress/block-hydration-experiments/pull/41 🎉🎉

I'm going to resolve the conflicts with https://github.com/WordPress/block-hydration-experiments/pull/28.

**Comment by luisherranz** [comment]
> I'm going to resolve the conflicts with #28.

Done. I think everything should be ready now.

**Comment by michalczaplinski** [comment]
- https://github.com/WordPress/block-hydration-experiments/pull/28 has been merged. 

As Luis mentioned in [this comment](https://github.com/WordPress/block-hydration-experiments/pull/28#issuecomment-1194147957) already, the approach of using the MutationObserver did not work so we're still going to wrap the non-interactive blocks with a custom element. However, Luis has moved the serialization to PHP so we're not saving the context in the content in `save()` anymore.

**Comment by luisherranz** [comment]
Thanks, Michal 🙂

I'll work on https://github.com/WordPress/block-hydration-experiments/issues/37 today, and if I can, I'll check if https://github.com/WordPress/block-hydration-experiments/issues/29 can be solved during `render_block`.

**Comment by luisherranz** [comment]
I've opened an issue to talk about `prettier-php`. If you're playing with this repository, please help us test it and if you have any problem, please report it here:

- https://github.com/WordPress/block-hydration-experiments/issues/44

**Comment by luisherranz** [comment]
The PR for the nomenclature changes is ready for review:

- https://github.com/WordPress/block-hydration-experiments/pull/45

**Comment by luisherranz** [comment]
## Update 3

Ok, now that we've kind of reached another milestone, I did another video summary to explain the latest changes.

<a href="https://www.loom.com/share/3b68a95ac80f43ecbe86ca8225cb35a0">
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/3b68a95ac80f43ecbe86ca8225cb35a0-with-play.gif">
  </a>

https://www.loom.com/share/3b68a95ac80f43ecbe86ca8225cb35a0

**Comment by luisherranz** [comment]
I've opened a new issue for the bug I discovered during the video: 
- https://github.com/WordPress/block-hydration-experiments/issues/46

**Comment by michalczaplinski** [comment]
> The PR for the nomenclature changes is ready for review:
> 
> * [Nomenclature changes #45](https://github.com/WordPress/block-hydration-experiments/pull/45)

I've merged this PR as well! 🙂

**Comment by SantosGuillamot** [comment]
I've created [this repo](https://github.com/SantosGuillamot/bhe-alpine/) to replicate the experiments done here but using Alpinejs. The main idea is to be able to compare the developer experience and limitations of both approaches. Right now, I just recreated the current blocks, but we will probably run a few more tests there.

**Comment by DAreRodz** [comment]
I took the liberty of merging the fix for the `WpBlock` unmounting issue as it was a simple PR and was tested already.

- https://github.com/WordPress/block-hydration-experiments/pull/47

I also updated the tracking issue description with the latest changes.

**Comment by luisherranz** [comment]
I've updated the Up Next list with two new items:

- Add support for frontend-side i18n
- Research how to hydrate different block variations

**Comment by ockham** [comment]
I've filed #51 (which Carlos approved and merged -- thanks!) to fix

- #46

**Comment by luisherranz** [comment]
I've opened the issue to make sure the order of execution of the `registerBlockType()` and `connectedCallback` is not important:

- https://github.com/WordPress/block-hydration-experiments/issues/52

**Comment by luisherranz** [comment]
@darerodz and have explored a fix for the race condition problem:

- https://github.com/WordPress/block-hydration-experiments/pull/57

Reviews are welcomed!

**Comment by DAreRodz** [comment]
I've discussed with @luisherranz what things we need to solve before exposing these APIs as experimental from Gutenberg.

It ended in a to-do list, included in the [issue description](https://github.com/WordPress/block-hydration-experiments/issues/39#issue-1289769123). ☝️

**Comment by luisherranz** [comment]
I've added a bit more context to the _Export different code depending on the context_ issue.

- https://github.com/WordPress/block-hydration-experiments/issues/11#issuecomment-1226878946

**Comment by DAreRodz** [comment]
I created a new issue for the remaining item in the "Required for experimental API" list.

- https://github.com/WordPress/block-hydration-experiments/issues/59

**Comment by luisherranz** [comment]
I've added a proposal for an alternative wrapperless hydration method based on a static virtual DOM of the full page:
- https://github.com/WordPress/block-hydration-experiments/issues/60

**Comment by luisherranz** [comment]
We opened a new issue to discuss possible solutions to locate the children of the islands/client-components (`wp-inner-blocks`/`slot`):

- https://github.com/WordPress/block-hydration-experiments/issues/62

**Comment by luisherranz** [comment]
@darerodz and I have created a new Tracking Issue to track the work related to the [Directives Hydration](https://github.com/WordPress/block-hydration-experiments/issues/60):

- https://github.com/WordPress/block-hydration-experiments/issues/64

From now on, please use this Tracking Issue for the Custom Elements Hydration experiment and the other one for the Directives Hydration experiment.

We've also renamed the `main` branch to `main-custom-elements-hydration` for this experiment and created a new one called `main-directives-hydration` for the other one. Please, select the correct branch when opening pull requests.

**Comment by SantosGuillamot** [comment]
- https://github.com/WordPress/block-hydration-experiments/issues/69

As explained in this issue, it looks like the [setTimeout callback](https://github.com/WordPress/block-hydration-experiments/blob/dynamic-blocks/src/gutenberg-packages/hydration.js#L128) we are using inside `connectedCallback` is not 100% reliable. So we may need to look for an alternative.

**Comment by luisherranz** [comment]
To avoid confusion, I've renamed "Island Hydration" to "Custom Elements Hydration" because both this and Directives are just ways of doing island hydration.

**Comment by luisherranz** [comment]
Closed as we're not actively working on this experiment anymore. Our work continues in:

- https://github.com/WordPress/block-hydration-experiments/issues/80

**Comment by fabiankaegy** [comment]
> Closed as we're not actively working on this experiment anymore. Our work continues in:

@luisherranz I've been trying to track down the reasons why this approach was abandoned :) Would you be able to shed some light on that? :) Thanks in advance!

**Comment by luisherranz** [comment]
It had a few issues:

- There's no easy way to use this approach with dynamic blocks.
- Any dynamic markup modifications (done with PHP hooks) are not reflected on the JSX template and are reverted after hydration.
- It also makes some things like internationalization much more complex than needed.

**Comment by fabiankaegy** [comment]
Thanks for sharing that insight :)
